lib/Sema/CSSolver.cpp

```diff
@@ -1858,60 +1858,44 @@ static Constraint *selectBestBindingDisjunction(
 
   // Collect any disjunctions that simply attempt bindings for a
   // type variable.
-  SmallVector<Constraint *, 8> bindingDisjunctions;
+  SmallVector<std::pair<Constraint *, TypeVariableType *>, 4>
+      bindingDisjunctions;
+
+  auto getAsTypeVar = [&cs](Type type) {
+    return cs.simplifyType(type)->getRValueType()->getAs<TypeVariableType>();
+  };
+
   for (auto *disjunction : disjunctions) {
-    TypeVariableType *commonTypeVariable = nullptr;
-    if (llvm::all_of(
-            disjunction->getNestedConstraints(),
-            [&](Constraint *bindingConstraint) {
-              if (bindingConstraint->getKind() != ConstraintKind::Bind)
-                return false;
-
-              auto *tv = cs.simplifyType(bindingConstraint->getFirstType())
-                             ->getRValueType()
-                             ->getAs<TypeVariableType>();
-              // Only do this for simple type variable bindings, not for
-              // bindings like: ($T1) -> $T2 bind String -> Int
-              if (!tv)
-                return false;
-
-              // If we've seen a variable before, make sure that this is
-              // the same one.
-              if (commonTypeVariable == tv)
-                return true;
-              if (commonTypeVariable)
-                return false;
-
-              commonTypeVariable = tv;
-              return true;
-            })) {
-      bindingDisjunctions.push_back(disjunction);
-    }
+    auto choices = disjunction->getNestedConstraints();
+    assert(!choices.empty());
+
+    auto *choice = choices.front();
+    if (choice->getKind() != ConstraintKind::Bind)
+      continue;
+
+    // We can judge disjunction based on the single choice
+    // because all of choices (of bind overload set) should
+    // have the same left-hand side.
+    // Only do this for simple type variable bindings, not for
+    // bindings like: ($T1) -> $T2 bind String -> Int
+    if (auto *typeVar = getAsTypeVar(choice->getFirstType()))
+      bindingDisjunctions.push_back({disjunction, typeVar});
```

Comment:
As mentioned in the other review thread we should be able to just immediately gather the constraints for the type variable and do the test below rather than collecting all the disjunctions and then going back through them and gathering constraints.

Reply:
Ah sure, I can do that too! :)

Reply:
Just didn't want to change the shape too much...

Reply:
Alright, new changes are in!